ph5/clients/ph5toms.py

```diff
@@ -511,32 +510,28 @@ def create_trace(self, station_to_cut, mp=False):
         for stc in station_to_cut_segments:
             das = self.ph5.query_das_t(stc.das, stc.component,
                                        stc.starttime,
-                                       stc.endtime,
-                                       stc.sample_rate,
-                                       stc.sample_rate_multiplier)
+                                       stc.endtime)
 
             if not das:
                 return
-
             das = [x for x in das]
             Das_tf = next(iter(das or []), None)
             if Das_tf is None:
                 return
             else:
                 das_t_start = (float(Das_tf['time/epoch_l']) +
                                float(Das_tf['time/micro_seconds_i']) / 1000000)
-
             if float(das_t_start) > float(stc.starttime):
                 start_time = das_t_start
 
             else:
                 start_time = stc.starttime
 
             nt = stc.notimecorrect
-
-            if stc.sample_rate > 0:
-                actual_sample_rate = float(
-                    stc.sample_rate) / float(stc.sample_rate_multiplier)
+            if Das_tf['sample_rate_i'] > 0:
+                actual_sample_rate = \
+                    float(Das_tf['sample_rate_i']) / \
+                    float(Das_tf['sample_rate_multiplier_i'])
```

Comment:
Commit 2565318b06e6e29244d0b561cc2fb801c4b01a61 addresses:
Removing the PH5 validate portion from the pull request. 

Adds a check between the DAS sample rate and Array sample rate as uses the array table sample rates as the actual sample rate:

```
            nt = stc.notimecorrect
            if Das_tf['sample_rate_i'] == stc.sample_rate:
                if Das_tf['sample_rate_i'] > 0:
                    actual_sample_rate = float(
                        stc.sample_rate) / float(stc.sample_rate_multiplier)
                else:
                    actual_sample_rate = 0
            else:
                continue
```
An additional test was also added with to ph5toms_test. 

> For ph5_availability, we need some test that really show the changes work. Since those functions always consider SEED channel, we need some das data that have the same SEED channels but different sample rate for testing. 

I may be missing something but it doesn't seem like the das tables have a field for SEED channel, SEED station or SEED Location. That is part of the challenge with coding this patch, there does not seem to be a robust foreign key that links the DAS and Array tables. 

DAS Example:
```
{'time/micro_seconds_i': 380000, 'sample_rate_i': 500, 'array_name_data_a': 'Data_a_0002', 'sample_count_i': 15000, 'stream_number_i': 1, 'time_table_n_i': 0, 'event_number_i': 2, 'array_name_SOH_a': '', 'response_table_n_i': 0, 'array_name_log_a': '', 'sample_rate_multiplier_i': 1, 'time/epoch_l': 1502294430, 'time/type_s': 'BOTH', 'raw_file_name_s': '3ch.fcnt', 'receiver_table_n_i': 1, 'array_name_event_a': '', 'time/ascii_s': 'Wed Aug  9 16:00:30 2017', 'channel_number_i': 1}
```
This seems to be consistent with the documentation. https://github.com/PIC-IRIS/PH5/wiki/Receivers-Group

Is there a different way we could think about this test?

Reply:
@timronan you are correct that das tables don't have SEED info. You have to define SEED info in array tables. The info that links DAS tables and array table includes: 
+ Array: das_serial_number_i, channel_number_i, sample_rate_i, sample_rate_multiplier_i, deploy and pickup time
+ Das: channel_number_i, sample_rate_i, sample_rate_multiplier_i, time and sample_count_i

Reply:
Based on testing:
```
    Array: das_serial_number_i, channel_number_i, sample_rate_i, sample_rate_multiplier_i, deploy and pickup time
    Das: channel_number_i, sample_rate_i, sample_rate_multiplier_i, time and sample_count_i

```
All seem to be non-unique identifiers, there are multiple DAS tables returned for each Array:das_serial_number_i. Each one of these DAS tables may or may not have matching channel number, sample rate identifiers, sample_rates, time and sample_count_i correspounding to deploy and pickup times. Different sample rates can lead to different count numbers and different Endtimes.  

This pull request is based on the issue 424 which describes that the DAS and Array tables have inconsistent sample rates so no data is returned. 

We could make a test similar to the test_ph5toms.

```
   def test_create_trace(self):

        ph5API_object = ph5api.PH5(
            path=os.path.join(os.getcwd(), 'ph5/test_data/ph5'),
            nickname='master.ph5')
        ph5toms = PH5toMSeed(ph5API_object)
        ph5toms.process_all()
        cuts = ph5toms.create_cut_list()
        for cut in cuts:
            trace = ph5toms.create_trace(cut)
            if trace is not None:
                self.assertEqual(trace[0].stats.sampling_rate, cut.sample_rate)
```

Also, the PH5Availabilty patch was previously tested on experiment 04-24 which is how the issue was originally discovered. The metadata for experiment 04-24 has been changed. 

The original test_ph5availability.py code provides test coverage for this patch. The two arrays refer to the same DAS serial number but have different sample rates. That is exactly what this pull request is trying to isolate and fix. 

```
/Experiment_g/Sorts_g/Array_t_002
	id_s=0407
	deploy_time/epoch_l=1545085230
	deploy_time/micro_seconds_i=0
	pickup_time/epoch_l=1547489524
	pickup_time/micro_seconds_i=0
	das/serial_number_s=5553
	seed_band_code_s=H
	sample_rate_i=200
	sample_rate_multiplier_i=1
	seed_instrument_code_s=H
	seed_orientation_code_s=N

/Experiment_g/Sorts_g/Array_t_003
	id_s=0407
	deploy_time/epoch_l=1545085230
	deploy_time/micro_seconds_i=0
	pickup_time/epoch_l=1547489524
	pickup_time/micro_seconds_i=0
	das/serial_number_s=5553
	seed_band_code_s=L
	sample_rate_i=100
	sample_rate_multiplier_i=1
	seed_instrument_code_s=H
	seed_orientation_code_s=N
```
Does this seems satisfactory?

Reply:
It is supposed to be unique group identifier for the correct one. The following is how to identify which time slots in das table match with a station entry in array table.
+ From Array table's station entry, you identify with das table to look at by das serial number. 
+ Then match channel_number_i and sample rate with the time slots in das table. 
+ Check if the time slot (time and sample count) is in the time range defined by station entry's deploy time and pickup time. One station entry may have multi das's time slots. If a station entry have no das tabel's time slot for it or a time slot not belong to a station entry, it is suppose to be report when running ph5validate.

Go back to the data needed to test your change, you need to create array table with 2 station entries with same SEED channel info, same channel_number_i but different deploy/pickup time and sample rate. You also need das data that match with those entries. Those are the kind of data the old code cannot pick up before and your code improve that. Because all the current tests are passed by old code, we need to point out what differences the changes make.

Reply:
The problem is that there can be multiple DAS tables with the same serial number that can or cannot have consistent sample rates, sample counts,  start_epoch and stop_epoch. To get DAS tables using array table information one must use the function ph5api.query_das_t, which only requires DAS serial number as an input argument. If start_epoch and/or stop_epoch are not supplied as arguments to ph5api.query_das_t, then the function looks for matching DAS serial numbers through all of time. Unless we want to limit the functionality of ph5api.query_das_t, and likely refactor some code, we cannot make a time based solution. Even if we changed ph5api.query_das_t to require start_epoch and stop_epochs than the solution could still be none unique because PH5 allows DAS tables with consistent serial numbers to have inconsistent sample rates and the check described above could fail. 

I added a test that is an exact match of the failing cases in the data repository. Refer to:
PH5/ph5/test_data/samplerate/
The new test cases are:
test_ph5_availability.test_check_samplerate
test_ph5toms.test_mismatch_sample_rate

I updated ph5toms and ph5availabilty to compare the sample rates between the DAS and array tables, if they match we use this sample rate and proceed. 

If the array tables sample rate does not match any of the returned DAS table sample rates, than we compare all of the returned to DAS table sample rates to themselves. If all of the DAS tables have the same sample rate we assume the metadata is incorrect  and use the DAS table sample rate.

If there are inconsistent sample rates in the DAS tables and they don't match the Array table sapmle rates than we throw an error and the metadata or data must be manually corrected by the PI. This is the logic:

```
                        Das_t = self.ph5.query_das_t(
                            ph5_das,
                            chan=ph5_channum,
                            start_epoch=ph5_start_epoch,
                            stop_epoch=ph5_stop_epoch)
                        for das in Das_t:
                            # Does Array.sr == DAS.sr? If so use sr
                            if das['sample_rate_i'] == st['sample_rate_i']:
                                samplerate_return = das['sample_rate_i']
                                ph5_sr = das['sample_rate_i']
                                times = self.ph5.get_availability(ph5_das,
                                                                  ph5_sr,
                                                                  ph5_channum,
                                                                  starttime,
                                                                  endtime)
                                empty_times = False
                            else:
                                continue
                        if empty_times is True:
                            for i, das in enumerate(Das_t):
                                # IF DAS.SR != Array.SR, USe DAS.SR if match
                                # Checks to see if all DAS tables have same SR
                                sr_prev = Das_t[i-1]['sample_rate_i']
                                if das['sample_rate_i'] != sr_prev:
                                    sr_mismatch = True
                            if sr_mismatch is True:
                                # Else throw warning and fail
                                LOGGER.error('DAS and Array Table sample' +
                                             ' rates do not match, DAS table' +
                                             ' sample rates do not match.' +
                                             ' Data must be updated.')
                            else:
                                # Uses SR if consistent
                                samplerate_return = das['sample_rate_i']
                                ph5_sr = das['sample_rate_i']
                                times = self.ph5.get_availability(ph5_das,
                                                                  ph5_sr,
                                                                  ph5_channum,
                                                                  starttime,
                                                                  endtime)
```
This logic passes all test cases. Master will fail the test cases:

test_ph5_availability.test_check_samplerate
test_ph5toms.test_mismatch_sample_rate